server/pkg/cache/cache.go

```diff
@@ -1,303 +1,557 @@
+			return fmt.Errorf("file %s not found in cache for website %s", resourceName, websiteAddress)
+		}
+		if err != nil {
+			return err
+		}
 
-	for _, f := range r.File {
-		if f.Name == resourceName {
-			rc, err := f.Open()
-			if err != nil {
-				return nil, fmt.Errorf("failed to open file %s in zip: %v", f.Name, err)
-			}
+		content, err = item.ValueCopy(nil)
+		if err != nil {
+			return err
+		}
 
-			defer closeReader(rc)
+		// Get the timestamp
+		item, err = txn.Get(timestampKey)
+		if err == badger.ErrKeyNotFound {
+			return fmt.Errorf("timestamp not found for file %s in website %s", resourceName, websiteAddress)
+		}
+		if err != nil {
+			return err
+		}
 
-			buf := new(bytes.Buffer)
+		timestamp, err = item.ValueCopy(nil)
+		if err != nil {
+			return err
+		}
 
-			_, err = io.Copy(buf, rc)
-			if err != nil {
-				return nil, fmt.Errorf("failed to copy content of file %s: %v", f.Name, err)
-			}
+		// Remove from disk
+		return c.deleteEntryFromDisk(txn, diskKey)
+	})
 
-			return buf.Bytes(), nil
-		}
+	if err != nil {
+		return nil, err
 	}
 
-	return nil, fmt.Errorf("file %s not found in archive %s", resourceName, archivePath)
-}
+	// Create a cache entry
+	entry := &cacheEntry{
+		content:        content,
+		modified:       time.Unix(0, int64(binary.BigEndian.Uint64(timestamp))),
+		websiteAddress: websiteAddress,
+		resourceName:   resourceName,
+	}
 
-// Save a resource in the cache for a given website
-func (c *Cache) Save(websiteAddress string, resourceName string, content []byte) error {
-	archivePath := c.getArchivePath(websiteAddress)
-	fileLock := c.newFileLock(archivePath)
-	tempArchivePath := archivePath + ".tmp"
+	// Add to RAM cache - eviction will be handled automatically by the callback
+	c.ramCache.Add(key, entry)
 
-	err := writeLockFile(fileLock)
+	return content, nil
+}
+
+// deleteEntryFromDisk removes an entry from the disk cache
+func (c *Cache) deleteEntryFromDisk(txn *badger.Txn, key []byte) error {
+	// Create the entry key prefix
+	entryKeyPrefix := make([]byte, 1+len(key))
+	entryKeyPrefix[0] = entryTag
+	copy(entryKeyPrefix[1:], key)
+
+	// Get the ID counter for this entry
+	idCounterKey := make([]byte, len(entryKeyPrefix)+1)
+	copy(idCounterKey, entryKeyPrefix)
+	idCounterKey[len(entryKeyPrefix)] = entrySubTagID
+
+	item, err := txn.Get(idCounterKey)
+	if err == badger.ErrKeyNotFound {
+		return nil // Entry not found in disk cache
+	}
 	if err != nil {
-		//nolint:wrapcheck
 		return err
 	}
 
-	defer lockUnlock(fileLock)
-
-	var buf bytes.Buffer
-	w := zip.NewWriter(&buf)
-
-	err = c.copyExistingEntries(archivePath, w, resourceName)
+	idCounterValue, err := item.ValueCopy(nil)
 	if err != nil {
 		return err
 	}
 
-	header := &zip.FileHeader{
-		Name:     resourceName,
-		Method:   zip.Deflate,
-		Modified: time.Now(),
-	}
+	// Create data and timestamp keys
+	dataKey := make([]byte, len(entryKeyPrefix)+1)
+	copy(dataKey, entryKeyPrefix)
+	dataKey[len(entryKeyPrefix)] = entrySubTagData
```

Comment:
you can replace these 3 lines by:
`dataKey := append(entryKeyPrefix, entrySubTagData)`

It should be true for all other similar code in your PR

Reply:
tried it, caused crashes and memory corruption in Badger transactions as it requires immutable underlying memory for variables within transactions and under some circumstances, go optimizes things a bit by rearranging memory on append()

Reply:
added comments in the key construction functions to clarify that bit

Reply:
more info here https://github.com/hypermodeinc/badger/blob/62c6dc2af0819cab2f1ad4e7a7e9adb07a640f89/txn.go#L401-L402